aleph2_logging_service/src/com/ikanow/aleph2/logging/utils/LoggingUtils.java

```diff
@@ -0,0 +1,147 @@
+	 * 
+	 * @param subsystem
+	 * @return
+	 */
+	public static DataBucketBean getExternalBucket(final String subsystem, final Level default_log_level) {
+		return BeanTemplateUtils.build(DataBucketBean.class)
+				.with(DataBucketBean::full_name, EXTERNAL_PREFIX + "/" + subsystem + "/")	
+				.with(DataBucketBean::management_schema, BeanTemplateUtils.build(ManagementSchemaBean.class)
+							.with(ManagementSchemaBean::logging_schema, BeanTemplateUtils.build(LoggingSchemaBean.class)
+										.with(LoggingSchemaBean::log_level, default_log_level)
+									.done().get())
+						.done().get())
+				.done().get();
+	}
+	
+
+	
+	/**
+	 * Returns back a IDataWriteService pointed at a logging output location for the given bucket
+	 * @param bucket
+	 * @return
+	 */
+	public static MultiDataService getLoggingServiceForBucket(final IServiceContext service_context, final DataBucketBean bucket) {
+		//change the bucket.full_name to point to a logging location
+		final DataBucketBean bucket_logging = BucketUtils.convertDataBucketBeanToLogging(bucket);
+		//return crudservice pointing to this path
+		return MultiDataService.getMultiWriter(bucket_logging, service_context); 
+	}
+
+	/**
+	 * Creates a map of subsystem -> logging level for quick lookups.  Grabs the overrides from
+	 * bucket.management_schema().logging_schema().log_level_overrides()
+	 * 
+	 * Returns an empty list if none exist there
+	 * 
+	 * @param bucket
+	 * @return
+	 */
+	public static ImmutableMap<String, Level> getBucketLoggingThresholds(final DataBucketBean bucket) {
+		//if overrides are set, create a map with them and the default
+		if (bucket.management_schema() != null &&
+				bucket.management_schema().logging_schema() != null ) {
+			return new ImmutableMap.Builder<String, Level>()
+		 	.put(DEFAULT_LEVEL_KEY, bucket.management_schema().logging_schema().log_level())
+		 	.putAll(Optional.ofNullable(bucket.management_schema().logging_schema().log_level_overrides()).orElse(new HashMap<String, Level>()))
+		 .build();
+		} else {
+			//otherwise just return an empty map
+			return new ImmutableMap.Builder<String, Level>().build();
+		}
+	}
+	
+	/**
+	 * Gets the minimal log level for the given subsystem by checking:
+	 * 1. if logging_overrides has that subsystem as a key
+	 * 2. if not, checks if there is a default level set
+	 * 3. if not, uses default_log_level
+	 * Then returns if the passed in log level is above the minimal log level or not.
+	 * 
+	 * @param level
+	 * @param logging_overrides
+	 * @param subsystem
+	 * @param default_log_level
+	 * @return
+	 */
+	public static boolean meetsLogLevelThreshold(final Level level, final ImmutableMap<String, Level> logging_overrides, final String subsystem, final Level default_log_level) {
+		final Level curr_min_level = 
+				Optional.ofNullable(logging_overrides.get(subsystem))
+				.orElse(Optional.ofNullable(logging_overrides.get(DEFAULT_LEVEL_KEY))
```

Comment:
This should be an orElseGet(() -> ...) i think? as it standard it does the extra lookup every time you enter the function even if not needed